pyogrio/_io.pyx

```diff
@@ -486,11 +530,28 @@ cdef apply_where_filter(OGRLayerH ogr_layer, str where):
     if err != OGRERR_NONE:
         try:
             exc_check()
-            name = OGR_L_GetName(ogr_layer)
         except CPLE_BaseError as exc:
             raise ValueError(str(exc))
 
-        raise ValueError(f"Invalid SQL query for layer '{name}': '{where}'")
+        raise ValueError(f"Invalid SQL query for layer '{OGR_L_GetName(ogr_layer)}': '{where}'")
+
+    # attempt to detect invalid SQL query for GPKG by checking count and first
```

Comment:
I had to add this section to catch invalid SQL errors for GPKG sooner; otherwise we were catching it while iterating over features later, which was not ideal because that code was far-removed from SQL / `where` related issues.

Reply:
Hmm... not sure, but if I understand correctly, when a correct where is specified this will result in all data being read + the where being applied 3 times:

1. once in `apply_where_filter`: to be able to count the features you need to process all data, or at least the data in the where
2. once in `validate_feature_range`: also to get the feature count
3. once when we actually get the data

The where filter only supports a quite limited subset of sql, so possibly applying the where isn't that heavy in the typical case... but still it sounds an "expensive" way to detect the error sooner.

Maybe the where-error-detecting logic can be moved to the new function `get_feature_count`? This way it will be triggered anyway before the loop fetching the data because it is called in `validate_feature_range` without adding extra processing compared to the current situation?

Reply:
True, that would at least avoid the double, potentially slow iteration within GDAL via `OGR_L_GetFeatureCount`.